### DISCUSSION TITLE
How to use render.py

### DISCUSSION THREAD

**Discussion by TITC** [lukas-blecher/LaTeX-OCR#136] [Q&A] [answered]
Here is my dataset [demo(200 equations)](https://drive.google.com/file/d/1EOmyj7mYjAk2SFudP6RDZ7AlY4oYQ-i5/view?usp=sharing), I call the file use `python render.py -i 200.lst  -o 200`.

The shell hasn't any error output, but nothing rendered at folder **200**. 

---
I wrote a small demo for render and it works fine with these data, but I think your code can make more variability for equations, such as font. so I wonder how to use it.

```python
import os
from tqdm.std import tqdm
from handytools import op
current_dir = op.parent_dir(__file__)
os.chdir(current_dir)
os.makedirs(os.path.join(current_dir, 'rendered'), exist_ok=True)
formulas = op.readlines("/mnt/d/data/NLP/wikipedia/latex_1024_200.lst")
formulas = [_ for _ in formulas if len(_) != 0]
for i, formula in enumerate(tqdm(formulas, total=len(formulas))):
    latex_code = """\\documentclass[preview]{standalone}\n\\begin{document}\n\\[ %s \\]\n\\end{document}""" % formula
    op.savelist(
        [latex_code], "%s/temp_simple.tex" % current_dir)
    os.system('latex2png -d 1080 %s/temp_simple.tex' % current_dir)
    os.system("cp -r %s/temp_simple.png %s/rendered/%s.png" %
              (current_dir, current_dir, str(i)))
```

**Comment by lukas-blecher** [comment, marked as answer]
The command worked fine for me. My guess is that your there are missing some packages for the latex to compile?
This is my template
```latex
\documentclass[varwidth]{standalone}
\usepackage{fontspec,unicode-math}
\usepackage[active,tightpage,displaymath,textmath]{preview}
\setmathfont{%s}
\begin{document}
\thispagestyle{empty}
%s
\end{document}
```
Try to compile it and see if it throws errors. You also need to install some of the fonts used `'GFSNeohellenicMath.otf', 'Asana Math', 'XITS Math', 'Cambria Math'`. These are all I could find online. I'd love to have even more though. Because I think this is one of the main reasons why this repo works much better with images found in the wild in comparison to similar projects. This and rendering the formulas with different DPIs.

Also what is the `latex2png` command you are using?

**Reply by TITC** [reply]
I have checked the fonts, at least `GFSNeohellenicMath.otf`, `Asana Math`, `XITS Math` is installed. 
```shell
(latex_ocr) yhtao@LG-gram:~$ find /usr/local/texlive/ -name GFSNeohellenicMath*
/usr/local/texlive/2022/texmf-dist/fonts/opentype/public/gfsneohellenicmath/GFSNeohellenicMath.otf
(latex_ocr) yhtao@LG-gram:~$ find /usr/local/texlive/ -name Asana*
/usr/local/texlive/2022/texmf-dist/fonts/opentype/public/asana-math/Asana-Math.otf
(latex_ocr) yhtao@LG-gram:~$ find /usr/local/texlive/ -name XITS*
/usr/local/texlive/2022/texmf-dist/fonts/opentype/public/xits/XITS-Bold.otf
/usr/local/texlive/2022/texmf-dist/fonts/opentype/public/xits/XITSMath-Regular.otf
/usr/local/texlive/2022/texmf-dist/fonts/opentype/public/xits/XITS-Regular.otf
/usr/local/texlive/2022/texmf-dist/fonts/opentype/public/xits/XITS-BoldItalic.otf
/usr/local/texlive/2022/texmf-dist/fonts/opentype/public/xits/XITS-Italic.otf
/usr/local/texlive/2022/texmf-dist/fonts/opentype/public/xits/XITSMath-Bold.otf
```

and incase random choiced a font not installed, I specified `GFSNeohellenicMath.otf`([change loop to 1](https://github.com/lukas-blecher/LaTeX-OCR/blob/59903b10d571deed49902d9ad5ff4da79a9b29c0/pix2tex/dataset/render.py#L108))
```shell
(latex_ocr) yhtao@LG-gram:/mnt/d$ python /mnt/d/git_repository/LaTeX-OCR/pix2tex/dataset/render.py -i /mnt/d/data/NLP/wikipedia/latex_1024_200.lst -o /mnt/d/data/NLP/wikipedia/latex_1024 -f GFSNeohellenicMath.otf
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:05<00:00,  1.85s/it]
/mnt/d/git_repository/LaTeX-OCR/pix2tex/dataset/render.py:127: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  same = names == prev_names
```
still nothing generated at output folder
```shell
(latex_ocr) yhtao@LG-gram:/mnt/d$ ll /mnt/d/data/NLP/wikipedia/latex_1024 
total 0
drwxrwxrwx 1 yhtao yhtao 512 Apr 24 15:10 ./
drwxrwxrwx 1 yhtao yhtao 512 Apr 24 16:36 ../
```

**Reply by lukas-blecher** [reply]
Why did you change it to 1 loop?
This might be the problem. If there is only one equation in each batch that is faulty for one reason or another no equations can be rendered.
Either change the batch size or increase the loops again.

**Reply by TITC** [reply]
about [latex2png](https://command-not-found.com/latex2png), and  [usage](https://man.archlinux.org/man/extra/latex2rtf/latex2png.1.en)

I changed batch size equal to 1 and loop 1, the input file contains 100 formulas, but nothing is generated. 
```shell
(latex_ocr) yhtao@LG-gram:~$ python /mnt/d/git_repository/LaTeX-OCR/pix2tex/dataset/render.py -i /mnt/d/data/CV/formula/im2latex-100k/im2latex_formulas.lst100.norm -o /mnt/d/data/NLP/wikipedia/latex_1024 -b 1 -f GFSNeohellenicMath.otf
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 101/101 [03:30<00:00,  2.08s/it]
/mnt/d/git_repository/LaTeX-OCR/pix2tex/dataset/render.py:131: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  same = names == prev_names
```


Thanks for your hint, the way I use it correct. Maybe something is wrong in my environment, I will debug tomorrow.

---
edit

because my laptop CPU is out of date, 12 loop is too slow for me in the test scenario.

**Reply by TITC** [reply]
but I think the script does not throw any errors is also a problem.

**Reply by TITC** [reply]
The reason of nothing output at my environment is we use different way to call `ImageMagick ` at [here](https://github.com/lukas-blecher/LaTeX-OCR/blob/59903b10d571deed49902d9ad5ff4da79a9b29c0/pix2tex/dataset/latex2png.py#L74)
```shell
            cmd = 'magick convert -density %i -colorspace gray %s -quality 90 %s' % (
                self.dpi,
                pdffile,
                pngfile,
            )  # -bg Transparent -z 9
```
in my environment, it is ` convert -density %i -colorspace gray %s -quality 90 %s`, I am not clear about the difference between them. but this [link](https://askubuntu.com/a/50173) same as mine.

![image](https://user-images.githubusercontent.com/35098797/165030141-9f7ed423-3e2f-4683-a72f-22a07587e583.png)

I install `apt install imagemagick` at server.
before install
```shell
host@server:~$ whereis convert
convert:

```

after install
```shell
host@server:~$ whereis convert
convert: /usr/bin/convert /usr/share/man/man1/convert.1.gz
```

**Reply by lukas-blecher** [reply]
> but I think the script does not throw any errors is also a problem.
I think throwing an error would be overkill but a `logging.error` message would be good, I agree.

Did you check if all packages in the template are installed on your machine?

**Reply by TITC** [reply]
now the script rendered normal, after some small tweak, will tell you after I tested.

---

follow [this manner](https://tex.stackexchange.com/a/248306) to check installed package

result list as below
```shell
(latex_ocr) yhtao@LG-gram:/mnt/d$ kpsewhich fontspec.sty
/usr/local/texlive/2022/texmf-dist/tex/latex/fontspec/fontspec.sty
(latex_ocr) yhtao@LG-gram:/mnt/d$ kpsewhich unicode-math.sty
/usr/local/texlive/2022/texmf-dist/tex/latex/unicode-math/unicode-math.sty
(latex_ocr) yhtao@LG-gram:/mnt/d$ kpsewhich active.sty
(latex_ocr) yhtao@LG-gram:/mnt/d$ kpsewhich tightpage.sty
(latex_ocr) yhtao@LG-gram:/mnt/d$ kpsewhich displaymath.sty
(latex_ocr) yhtao@LG-gram:/mnt/d$ kpsewhich textmath.sty
```
Does the output is normal? if abnormal, could you tell me how to install those missing package `active,tightpage,displaymath,textmath`.

**Reply by lukas-blecher** [reply]
So the thing is that `active,tightpage,displaymath,textmath` are only arguments for the package `preview`. That should return a path

**Reply by TITC** [reply]
installed all of them

```shell
(base) yhtao@LG-gram:~$ kpsewhich preview.sty
/usr/local/texlive/2022/texmf-dist/tex/latex/preview/preview.sty
(base) yhtao@LG-gram:~$ kpsewhich standalone.sty
/usr/local/texlive/2022/texmf-dist/tex/latex/standalone/standalone.sty
```

**Reply by TITC** [reply]
How do you know which font can be used to render? I found some font but has some exceptions.

The font I use is `STIXMath-Regular.otf`, it successfully rendered 53 formulas in the given 100 formulas.
```shell
Package fontspec Warning: Font "STIXMath-Regular" does not contain requested
(fontspec)                Script "Math".


Package unicode-math Warning: The first font loaded by unicode-math must be an
(unicode-math)                OpenType Math font (with script=math). If you
(unicode-math)                simply want ‘the default’ before loading
(unicode-math)                supplementary fonts over the top for certain
(unicode-math)                ranges, use: 
(unicode-math)                  \setmathfont{latinmodern-math.otf}


Package fontspec Warning: Font "STIXMath-Regular" does not contain requested
(fontspec)                Script "Math".


Package fontspec Warning: OpenType feature 'Style=MathScript' (ssty) not
(fontspec)                available for font 'STIXMath-Regular' with script
(fontspec)                'CustomDefault' and language 'Default'.


Package fontspec Warning: OpenType feature 'Style=MathScriptScript' (ssty) not
(fontspec)                available for font 'STIXMath-Regular' with script
(fontspec)                'CustomDefault' and language 'Default'.

(./eqcbu4tz91.aux)
Preview: Fontsize 10pt
Preview: PDFoutput 1
Preview: Tightpage -32891 -32891 32891 32891
[1] [2] [3] [4] [5] [6] [7] [8] [9] [10] [11] [12] [13] [14] [15] [16] [17]
[18] [19] [20] [21] [22] [23] [24] [25] [26] [27] [28] [29] [30] [31] [32]
[33] [34] [35] [36] [37] [38] [39] [40] [41] [42] [43] [44] [45]
Missing character: There is no 𝛤 (U+1D6E4) in font [lmroman10-bold]:mapping=tex
-text;!
Missing character: There is no 𝛤 (U+1D6E4) in font [lmroman10-bold]:mapping=tex
-text;!
[46] [47] [48] [49] [50] [51] [52]
./eqcbu4tz91.tex:60: Missing $ inserted.
```

**Reply by lukas-blecher** [reply]
I think the STIX font is equivalent with the XITS Math already in use.
> Khaled Hosny’s XITS and XITS Math are forks of STIX that fix its bugs, and have the same glyphs.
[source](https://tex.stackexchange.com/a/550872/269067)

But the output you shared only speaks of warnings. Did the same formulas work for another font? I feel like this should not have interrupted the rendering but I could be wrong.

**Reply by TITC** [reply]
You're right, `Latin Modern Math` and `GFSNeohellenicMath.otf` also not work. The render process is interrupted by `$$ G ( \textbf { r } , \textbf { r ^ { \prime } } ) $$`

```shell
./eqcbu4tz91.tex:60: Missing $ inserted.
<inserted text> 
                $
l.60 ...\textbf { r } , \textbf { r ^ { \prime } }
                                                   ) $$
./eqcbu4tz91.tex:60: Extra }, or forgotten $.
\textdef@ ...th {#1}\let \f@size #2\selectfont #3}
                                                  }
l.60 ...\textbf { r } , \textbf { r ^ { \prime } }
                                                   ) $$
```

**Reply by lukas-blecher** [reply]
I feel like that should be a `\mathbf` instead of `\textbf`.
Interesting. Could easily be replaced in every instance I belive. No need for textbf in math code
Edit: apparently spaces are not ignored in `\textbf`. So not so trivial after all

**Reply by TITC** [reply]
yeah, you are right. maybe it can be added to preprocessing.

**Reply by lukas-blecher** [reply]
While that's true, it should not even happen in the first place. As you noticed yourself the code wouldn't even compile. 
I think fixing latex code is too much for preprocessing. The sample in question should just be dropped.

**Reply by bruno-who-likes** [reply]
I tested both 'Latin Modern Math' and  'GFS Neohellenic Math' on some of your examples in 200.lst. They both worked.

**Reply by TITC** [reply]
sorry, it's my fault. This problem is terminated at https://github.com/lukas-blecher/LaTeX-OCR/discussions/136#discussioncomment-2623686. And the other part from https://github.com/lukas-blecher/LaTeX-OCR/discussions/136#discussioncomment-2643289 is extend content, maybe you noticed this equation I mentioned at https://github.com/lukas-blecher/LaTeX-OCR/discussions/136#discussioncomment-2645023 can't be found at 200.lst file.
```tex
$$ G ( \textbf { r } , \textbf { r ^ { \prime } } ) $$
```
We are not tested at the same file. And you can add this equation in the middle of 200.lst then test again. I think you can see the interrupt both 'Latin Modern Math' and 'GFS Neohellenic Math'.
